sources/web/datalab/polymer/components/datalab-terminal/datalab-terminal.ts

```diff
@@ -0,0 +1,140 @@
+
+      self._boundResizeHandler = self._resizeHandler.bind(self);
+      window.addEventListener('resize', self._boundResizeHandler, true);
+
+      // Get the first terminal instance by calling the Jupyter API. If none are returned,
+      // start a new one.
+      ApiManager.listTerminalsAsync()
+        .then((terminals: [JupyterTerminal]) => {
+          return terminals.length === 0 ? ApiManager.startTerminalAsync() : terminals[0];
+        })
+        .then((terminal: JupyterTerminal) => {
+          self._initTerminal(terminal.name);
+        });
+    });
+  }
+
+  /**
+   * Creates the front end terminal element and connects it to the backend Jupyter terminal.
+   * TODO: Consider adding a loading experience while the connection is made.
+   * @param terminalName name of the Jupyter terminal session to be used in websocket connection
+   */
+  _initTerminal(terminalName: string) {
+    // Clear any older terminal elements created.
+    this.$.theTerminal.innerHTML = '';
+    this._xterm = new Terminal();
+
+    // First, create the connection to the Jupyter terminal.
+    this._wsConnection = new WebSocket('ws://' +
+                                        window.location.host +
+                                        '/terminals/websocket/' +
+                                        terminalName);
+
+    this._wsConnection.onopen = () => {
+      // Front-end to Jupyter.
+      this._xterm.on('data', (data: MessageEvent) => {
+        this._wsConnection.send(JSON.stringify(['stdin', data]));
+      });
+
+      // Jupyter to front-end. Intercept disconnect messages to reset the terminal.
+      this._wsConnection.onmessage = (event: any) => {
+        const data = JSON.parse(event.data);
+        if (data[0] === 'stdout') {
+          this._xterm.write(data[1]);
+        } else if (data[0] === 'disconnect') {
+          // TODO: Consider making the reset manual, by adding a toolbar with a button, or a popup
+          // that shows when the 'disconnect' message is sent back.
+          this._initTerminal(terminalName);
+        }
+      };
+
+      // Now, create the front-end terminal.
+      this._xterm.open(this.$.theTerminal, true);
+      this._resizeHandler();
+    };
+  }
+
+  /**
+   * Called when the element is detached from the DOM. Cleans up event listeners.
+   */
+  disconnectedCallback() {
+    if (this._boundResizeHandler) {
+      window.removeEventListener('resize', this._boundResizeHandler);
+    }
+  }
+
+  /**
+   * On window resize, both front-end and Jupyter terminal instances need to be resized and
+   * kept in sync, otherwise line wrapping issues will happen.
+   */
+  _resizeHandler() {
+    if (this._xterm) {
+      const rows = this.$.theTerminal.clientHeight / this._charHeight;
+      const cols = this.$.theTerminal.clientWidth / this._charWidth;
+
+      this._xterm.resize(Math.floor(cols), Math.floor(rows));
+
+      if (this._wsConnection.readyState === 1) {
```

Comment:
Under what conditions would a resize happen and the connection not be ready? If that happens, would the Jupyter terminal size then not match the front-end terminal size?